firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/UpdateApkClientTest.java

```diff
@@ -0,0 +1,169 @@
+// limitations under the License.
+
+package com.google.firebase.appdistribution;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.when;
+import static org.robolectric.Shadows.shadowOf;
+
+import android.app.Activity;
+import androidx.test.core.app.ApplicationProvider;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.shadows.ShadowActivity;
+
+@RunWith(RobolectricTestRunner.class)
+public class UpdateApkClientTest {
+
+  private static final String TEST_API_KEY = "AIzaSyabcdefghijklmnopqrstuvwxyz1234567";
+  private static final String TEST_APP_ID_1 = "1:123456789:android:abcdef";
+  private static final String TEST_PROJECT_ID = "777777777777";
+  private static final String TEST_URL = "https://test-url";
+  private static final long TEST_FILE_LENGTH = 1000;
+  private static final int RESULT_OK = -1;
+  private static final int RESULT_CANCELED = 0;
+  private static final int RESULT_FAILED = 1;
+
+  private UpdateApkClient updateApkClient;
+  private TestActivity activity;
+  private ShadowActivity shadowActivity;
+  private CountDownLatch latch;
+  @Mock private File mockFile;
+
+  static class TestActivity extends Activity {}
+
+  @Before
+  public void setup() {
+
+    MockitoAnnotations.initMocks(this);
+
+    FirebaseApp.clearInstancesForTest();
+
+    FirebaseApp firebaseApp =
+        FirebaseApp.initializeApp(
+            ApplicationProvider.getApplicationContext(),
+            new FirebaseOptions.Builder()
+                .setApplicationId(TEST_APP_ID_1)
+                .setProjectId(TEST_PROJECT_ID)
+                .setApiKey(TEST_API_KEY)
+                .build());
+
+    activity = Robolectric.buildActivity(TestActivity.class).create().get();
+    shadowActivity = shadowOf(activity);
+
+    this.latch = new CountDownLatch(1);
+    when(mockFile.getPath()).thenReturn(TEST_URL);
+    when(mockFile.length()).thenReturn(TEST_FILE_LENGTH);
+
+    this.updateApkClient = Mockito.spy(new UpdateApkClient(firebaseApp));
+  }
+
+  @Test
+  public void updateApk_whenDownloadFails_setsNetworkError() throws Exception {
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
```

Comment:
Why does this fail? I don't see any explicit failure returned by mocks in the tests

Reply:
Fails due to the httpsConnection inputStream being null since the Test URL isn't a real download URL, going to mock the httpsconnection to make it explicit